predicate/solver/ast.py

```diff
@@ -1637,6 +1751,84 @@ def traverse(self):
         return self.fn_map
 
 
+class StringSetMapEquals(LogicMixin):
```

Comment:
This is somewhat related to this PR but mostly commenting due to the fact that this expands the reliance on this pattern. I'm growing somewhat concerned about our usage of combinators and operators that retain the full type information about what they operate on such as `StringSetMapEquals`. This has the potential to create a quadric explosion in needed classes with very high code repeatability. Is there any way to generalize this tree-like AST structure and include the equals logic via some method on `StringSetMap` that the regular `Eq` class can use instead. This also applies to the other highly type specialized classes in this PR such as `StringSetMapForEachKeyConstraint`. This may not seem like such a pain here but it becomes more and more annoying both in terms of code readability, maintenance and my work to integrate this into Teleport.

Reply:
I replaced `StringSetMapEquals` with more generic class that accepts operator as a parameter.